NodeProg/Infrastructure/NumberNode.cs

```diff
@@ -21,14 +22,129 @@ public NumberNode(long value)
                 .Min() ?? input.Length;
 
             var numberLength = stopMarkerIndex - index;
-            var numberString = input.Substring(index, numberLength);
 
-            long value;
-
-            if (Int64.TryParse(numberString.Trim(), out value))
-                return new NodeParseResult { Node = new NumberNode(value), StartPosition = index, Length = numberLength };
+            var value = TryParseInt64(input, index, numberLength);
+            if (value != null)
+                return new NodeParseResult { Node = new NumberNode(value.Value), StartPosition = index, Length = numberLength };
             return null;
+        }
+        private static long? TryParseInt64(
+            string input,
+            int startIndex,
+            int length,
+            bool allowLeadingWhitespace = true,
+            bool allowTrailingWhitespace = true)
+        {
+            if (length == 0)
+                return null;
+            var trimChars = new List<char>();
+            if (allowLeadingWhitespace)
+                trimChars.Add(' ');
+            if (allowTrailingWhitespace)
+                trimChars.Add('\t');
+
+            var digits = new char[] { '0', '1', '2', '3', '4', '5', '6', '7', '8', '9' };
+
+            var position = startIndex;
+            var minusFlag = false;
+            var value = new Int64();
+            for (; position < startIndex + length;)
+            {
+                if (input[position] == '+' || input[position] == '-')
+                {
+                    if (input[position] == '-')
+                        minusFlag = true;
+                    position++;
+                    break;
+                }
+                var trimFlag = false;
+                foreach (var tChar in trimChars)
+                {
+                    if (input[position] == tChar)
+                    {
+                        trimFlag = true;
+                        position++;
+                        break;
+                    }
+                }
+                if (trimFlag)
+                    continue;
+                break;
+            }
+
+            var int64s = new List<int>() { };
+
+            for (; position < startIndex + length;)
+            {
+                var digitFlag = false;
+                foreach (var digit in digits)
+                {
+                    if (input[position] == digit)
+                    {
+                        digitFlag = true;
+                        int64s.Add(int.Parse(digit.ToString()));
```

Comment:
you don't need to store digits in an array. you could just calculate result on the fly on each iteration.